blockchain/src/blockchain/mod.rs

```diff
@@ -10,3 +13,25 @@ pub mod slots;
 pub mod verify;
 pub mod wrappers;
 pub mod zkp_sync;
+
+/// A post-validation hook that can be registered with the blockchain to run after block validation.
+/// For `extend` and `rebranch` operations, this is called before a block is committed to the database.
+/// It thus provides a timing advantage.
+pub trait PostValidationHook {
+    /// Run the pre-commit hook.
+    fn post_validation(&self, block: Block, push_result: Result<PushResult, PushError>);
+}
+
+/// A dummy pre-commit hook that does nothing.
+impl PostValidationHook for () {
+    fn post_validation(&self, _block: Block, _push_result: Result<PushResult, PushError>) {}
+}
+
+impl<F: PostValidationHook> PostValidationHook for Option<F> {
+    fn post_validation(&self, block: Block, push_result: Result<PushResult, PushError>) {
+        match self {
+            Some(hook) => hook.post_validation(block, push_result),
+            None => {}
+        }
```

Comment:
This should work, I think.
```suggestion
        self.as_ref().inspect(|hook| hook.post_validation(block, push_result));
```